### PR TITLE
refactor kube data client

### PR DESCRIPTION
**Pull request by zeitlinger**
split the methods in the kube data client

### CODE REVIEW

**Comment by tkrop** [comment]
:+1:

File: dataclients/kubernetes/kube.go [hunk 1]
```diff
@@ -802,217 +802,275 @@ func (c *Client) ingressToRoutes(state *clusterState, defaultFilters map[resourc
...
-		if c.kubernetesEnableEastWest {
-			for _, rule := range i.Spec.Rules {
-				if rs, ok := hostRoutes[rule.Host]; ok {
-					rs = append(rs, createEastWestRoutes(c.eastWestDomainRegexpPostfix, i.Metadata.Name, i.Metadata.Namespace, rs)...)
-					hostRoutes[rule.Host] = rs
-				}
+func (c *Client) addSpecRule(ru *rule, logger *log.Entry, backendWeights map[string]float64,
+	extraRoutes []*eskip.Route, i *ingressItem, pathMode PathMode, hostRoutes map[string][]*eskip.Route,
+	redirect *redirectInfo, state *clusterState, annotationFilter string,
+	defaultFilters map[resourceId]string, annotationPredicate string) error {
```
Comment (szuecs) [inline dataclients/kubernetes/kube.go]: 4 Lines of parameters?
Looks like a smell to me.

**Comment by zeitlinger** [comment]
What do you think about parameter structs that group related parameters?
I wasn't sure if this ideomatic.

**Comment by szuecs** [comment]
I am not sure if it makes it better. Normally the data should be on the object itself and only a few parameters. Maybe you can identify an object that covers most of it and probably other methods would be also of this receiver type and get reduced number of parameters, too.
How would you suggest to group them?

**Comment by zeitlinger** [comment]
@szuecs I've gave it a shot with the new `ingressContext` struct. Please tell me what you think.

File: dataclients/kubernetes/kube.go [hunk 2]
```diff
@@ -802,217 +817,258 @@ func (c *Client) ingressToRoutes(state *clusterState, defaultFilters map[resourc
...
+	}
 
-		// parse predicate annotation
-		var annotationPredicate string
-		if val, ok := i.Metadata.Annotations[skipperpredicateAnnotationKey]; ok {
-			annotationPredicate = val
+	return routes, nil
+}
+
+func (c *Client) ingressRoute(i *ingressItem, redirect *redirectInfo, state *clusterState,
```
Comment (aryszka) [inline dataclients/kubernetes/kube.go]: while extracting this block to a method looks safe, it still could be cleaner. What makes a bit hard to comprehend is that it returns one route, but potentially it creates more routes which are added to the hostRoutes argument as a side effect.

**Comment by aryszka** [comment]
:+1:

**Comment by szuecs** [comment]
:+1:

**Comment by szuecs** [comment]
Thanks for the effort @zeitlinger

**Pull request merged**